osrf/vrx
# Use a global synchronized wavefield

Fixes #564 using Gazebo Transport.

Before this patch, all the plugins using a wave field needed to have the same `<wavefield>` SDF parameter block in its SDF plugin definition. This pull request changes this behavior by:

* Setting one of the plugins to operate as the `primary`. The `primary` plugin requires to specify all the `<wavefield>` parameters as usual. Then, it should periodically publish the wavefield parameters via Transport.
* The rest of the plugins should be set to operate as `secondary`. This means passing only one parameters inside the `<wavefield>` block. This parameters is the topic name where the wave field parameters are sent. When received, the wavefield can be populated with a new helper function.

Our `WaveVisual` plugin is set to operate as the primary. All the instances of `Surface` and `PolyhedreaBuoyancyDrag` plugins are set to operate as secondary.

### How to test it?

* You can run the simulation as usual:
```
ros2 launch vrx_gz competition.launch.py world:=stationkeeping_task
```

and verify that the buoys and the WAM-V move a bit up and down.

* You can modify `vrx_gz/worlds/stationkeeping_task.sdf`:
```
    
    <plugin filename="libPublisherPlugin.so" name="vrx::PublisherPlugin">
      <message type="gz.msgs.Param" topic="/vrx/wavefield/parameters"
               every="2.0">
...
        params {
          key: "gain"
          value {
            type: DOUBLE
            double_value: 0.1
          }
        }
...
      </message>
    </plugin>
```

Recompile and launch the simulation again. You should verify now that the WAM-V and the buoys stay flat at the water surface.
[Edit: updated test instructions]
